File: lina_cpp/src/lina_cpp/llowfsc.py
"""lina_cpp.llowfsc - C++-backed math kernels + Python loop orchestration.

The math hot path of low-order WFS&C lives in ``cpp/src/llowfsc.cpp``:
* ``acquire_ref`` -- camera reference image (dark subtract, mask, flux normalize)
* ``reconstruct`` -- delta-image projection onto a control-matrix slice
* ``compute_zpo`` -- DM-command-driven zero-path-offset image accumulation
* ``loop_step``   -- one full closed-loop math iteration (returns delta DM cmd)

The closed-loop ``run()`` function stays in Python -- it orchestrates
camera reads, DM writes, and per-iteration logging via callables passed
in by the caller. This is the same orchestration as
``lina.llowfsc.run`` and is reused unchanged; only the math primitives
swap to the C++ implementations.

When you are ready to migrate this loop into a MagAOX C++ app, the
kernels in ``cpp/src/llowfsc.cpp`` are already written to be called
directly without Python in the loop -- you wire shmim reads/writes to
``lina::llowfsc::loop_step`` in the MagAOX app and you have a fully
native real-time controller.
"""

from __future__ import annotations

import copy
import time

import numpy as np

from . import _core
from .math_module import ensure_np_array


# ---------------------------------------------------------------------------
# Math kernels (delegate to C++)
# ---------------------------------------------------------------------------

def acquire_ref(take_im_fun, take_im_params, wfs_mask, camlo_dark=0.0, flux_norm=True):
    """Acquire a reference image: take a frame, dark-subtract, mask, normalize.

    Matches lina.llowfsc.acquire_ref signature. The math (subtract,
    mask multiply, sum, divide) runs in C++ via ``_core.llowfsc_acquire_ref``;
    the camera read goes through the user-supplied ``take_im_fun``.
    """
    camlo_ref_im = take_im_fun(**take_im_params)
    camlo_ref_im = ensure_np_array(camlo_ref_im).astype(np.float64, copy=False)
    mask = ensure_np_array(wfs_mask).astype(np.uint8, copy=False)

    # Dark can be a scalar OR an image; pass through unchanged.
    if isinstance(camlo_dark, (int, float)):
        dark_arg = float(camlo_dark)
    else:
        dark_arg = ensure_np_array(camlo_dark).astype(np.float64, copy=False)

    ref_image, flux_norm_coeff = _core.llowfsc_acquire_ref(
        camlo_ref_im, mask, dark_arg, bool(flux_norm)
    )
    if flux_norm:
        return ref_image, flux_norm_coeff
    return ref_image


def reconstruct(
    camlo_im,
    ref_im,
    wfs_mask,
    control_matrix,
    dark_im=0.0,
    modes=(0, 10),
    flux_norm=True,
    return_del_im=False,
):
    """Project a delta-image onto the [modes[0], modes[1]) rows of `control_matrix`.

    Matches lina.llowfsc.reconstruct signature. All math in C++.
    """
    camlo_im = ensure_np_array(camlo_im).astype(np.float64, copy=False)
    ref_im = ensure_np_array(ref_im).astype(np.float64, copy=False)
    mask = ensure_np_array(wfs_mask).astype(np.uint8, copy=False)
    C = ensure_np_array(control_matrix).astype(np.float64, copy=False)

    if isinstance(dark_im, (int, float)):
        dark_arg = float(dark_im)
    else:
        dark_arg = ensure_np_array(dark_im).astype(np.float64, copy=False)

    return _core.llowfsc_reconstruct(
        camlo_im, ref_im, mask, C,
        mode_lo=int(modes[0]), mode_hi=int(modes[1]),
        dark_im=dark_arg, flux_norm=bool(flux_norm),
        return_del_im=bool(return_del_im),
    )


def compute_zpo(
    DM_STREAMS,
    dm_mask,
    wfs_mask,
    response_matrix,
    dm_modal_matrix,
    ZPO_STREAM,
):
    """Accumulate ZPO image from DM command streams. Matches lina.llowfsc.compute_zpo.

    Each entry of ``DM_STREAMS`` is expected to expose a ``grab_latest()``
    method returning the current DM command as a 2D array; we extract
    the masked entries and pass them to the C++ kernel as a list of
    1D vectors.
    """
    dm_mask = ensure_np_array(dm_mask).astype(bool)
    wfs_mask_u8 = ensure_np_array(wfs_mask).astype(np.uint8, copy=False)
    R = ensure_np_array(response_matrix).astype(np.float64, copy=False)
    M = ensure_np_array(dm_modal_matrix).astype(np.float64, copy=False)

    cmds_masked = []
    for stream in DM_STREAMS:
        cmd = ensure_np_array(stream.grab_latest()).astype(np.float64, copy=False)
        cmds_masked.append(np.ascontiguousarray(cmd[dm_mask]))

    zpo = _core.llowfsc_compute_zpo(cmds_masked, wfs_mask_u8, R, M)
    if ZPO_STREAM is not None:
        ZPO_STREAM.write(zpo)
    return zpo


def loop_step(
    camlo_im,
    ref_plus_zpo,
    wfs_mask,
    control_matrix,
    dm_modes,
    gains,
    modes=(0, 10),
    ffo=None,
    dark_im=0.0,
    flux_norm=True,
):
    """One full per-iteration math step. Returns the delta DM command (2D).

    `dm_modes` may be either:
      - a (Nmodes, dm_rows, dm_cols) 3D array (lina convention), OR
      - a (Nmodes, dm_rows*dm_cols) 2D array (already flattened).

    `gains` is indexed relative to mode_lo (so gains[0] applies to
    mode `modes[0]`). `ffo` defaults to zero if None.
    """
    camlo_im = ensure_np_array(camlo_im).astype(np.float64, copy=False)
    ref_plus_zpo = ensure_np_array(ref_plus_zpo).astype(np.float64, copy=False)
    mask = ensure_np_array(wfs_mask).astype(np.uint8, copy=False)
    C = ensure_np_array(control_matrix).astype(np.float64, copy=False)
    M = ensure_np_array(dm_modes).astype(np.float64, copy=False)

    if M.ndim == 3:
        Nmodes, dm_rows, dm_cols = M.shape
        M_flat = M.reshape(Nmodes, dm_rows * dm_cols)
    else:
        Nmodes, Ndm = M.shape
        # We can't recover dm_rows/dm_cols from a 2D modes array; assume square.
        side = int(np.sqrt(Ndm))
        if side * side != Ndm:
            raise ValueError(
                "loop_step: cannot infer DM shape from flat modes; "
                "pass dm_modes as a 3D (Nmodes, rows, cols) array instead."
            )
        dm_rows = dm_cols = side
        M_flat = np.ascontiguousarray(M)

    g = ensure_np_array(gains).astype(np.float64, copy=False)
    nrange = int(modes[1]) - int(modes[0])
    if g.size == 1:
        g = np.full(nrange, float(g))
    if ffo is None:
        f = np.zeros(nrange)
    else:
        f = ensure_np_array(ffo).astype(np.float64, copy=False)

    if isinstance(dark_im, (int, float)):
        dark_arg = float(dark_im)
    else:
        dark_arg = ensure_np_array(dark_im).astype(np.float64, copy=False)

    return _core.llowfsc_loop_step(
        camlo_im, ref_plus_zpo, mask, C, np.ascontiguousarray(M_flat),
        int(dm_rows), int(dm_cols), g,
        int(modes[0]), int(modes[1]), f,
        dark_im=dark_arg, flux_norm=bool(flux_norm),
    )


# ---------------------------------------------------------------------------
# Closed-loop run() -- Python orchestration that calls the C++ kernels.
#
# Signature kept identical to lina.llowfsc.run so user code (notebooks,
# scripts) doesn't need to change. The math hot path now goes through
# the C++ extension via loop_step / reconstruct.
# ---------------------------------------------------------------------------

def run(
    take_im_fun,
    take_im_params,
    set_dm_fun,
    set_dm_params,
    get_dm_fun,
    get_dm_params,
    get_gains,
    ref_im,
    control_matrix,
    dm_modes,
    wfs_mask,
    dark_im=0.0,
    get_zpo=None,
    get_zpo_params=None,
    get_ffo=None,
    get_ffo_params=None,
):
    """One iteration of the closed loop. Use in a TimedThread / loop driver.

    Identical to lina.llowfsc.run but with the math hot path delegated
    to C++ via loop_step. Math: same. Signature: same. Behaviour: same.
    """
    if get_zpo_params is None: get_zpo_params = {}
    if get_ffo_params is None: get_ffo_params = {}

    camlo_im = take_im_fun(**take_im_params)

    zpo = get_zpo(**get_zpo_params) if get_zpo is not None else 0.0
    ref_plus_zpo = ref_im + zpo

    ffo = get_ffo(**get_ffo_params) if get_ffo is not None else None

    gains = get_gains()
    # The Python lina.llowfsc.run hardcodes modes=(0,10); preserve that.
    modes = (0, 10)
    del_dm_command = loop_step(
        camlo_im, ref_plus_zpo, wfs_mask, control_matrix, dm_modes,
        gains, modes=modes, ffo=ffo, dark_im=dark_im, flux_norm=True,
    )

    total_dm_command = get_dm_fun(**get_dm_params) + del_dm_command
    set_dm_fun(total_dm_command, **set_dm_params)


# ---------------------------------------------------------------------------
# calibrate_dm_modes still uses the user-supplied callables for camera /
# DM I/O; the math is just trivial differencing and dividing. We delegate
# to lina.llowfsc.calibrate_dm_modes (it's a thin loop, not a hot path).
# ---------------------------------------------------------------------------

from lina.llowfsc import (  # noqa: E402, F401
    ContinuousProcess,
    TimedThread,
    calibrate_dm_modes,
    plot_responses,
)

File: cpp/src/llowfsc.cpp
#include "lina/llowfsc.h"

#include "lina/linalg.h"

#include <algorithm>
#include <cmath>
#include <cstring>
#include <stdexcept>

namespace lina::llowfsc {

namespace {

// Common shape check used everywhere in this file.
void require_same_shape(const char* who,
                        std::size_t a_rows, std::size_t a_cols,
                        std::size_t b_rows, std::size_t b_cols) {
    if (a_rows != b_rows || a_cols != b_cols) {
        throw std::invalid_argument(
            std::string(who) + ": shape mismatch ("
            + std::to_string(a_rows) + "x" + std::to_string(a_cols) + " vs "
            + std::to_string(b_rows) + "x" + std::to_string(b_cols) + ")");
    }
}

// Sum of `arr` restricted to entries where mask != 0.
double sum_masked(const Array2D<double>& arr,
                  const Array2D<std::uint8_t>& mask) {
    double s = 0.0;
    const std::size_t n = arr.size();
    const double* p = arr.data();
    const std::uint8_t* m = mask.data();
    for (std::size_t i = 0; i < n; ++i) {
        if (m[i]) s += p[i];
    }
    return s;
}

// Count of mask entries that are non-zero.
std::size_t count_mask(const Array2D<std::uint8_t>& mask) {
    std::size_t c = 0;
    const std::size_t n = mask.size();
    const std::uint8_t* m = mask.data();
    for (std::size_t i = 0; i < n; ++i) {
        if (m[i]) ++c;
    }
    return c;
}

// Extract `arr[mask]` into a flat std::vector<double>. The order
// matches NumPy's row-major iteration (the same as lina.llowfsc).
std::vector<double> extract_masked(const Array2D<double>& arr,
                                   const Array2D<std::uint8_t>& mask) {
    std::vector<double> out;
    out.reserve(count_mask(mask));
    const std::size_t n = arr.size();
    const double* p = arr.data();
    const std::uint8_t* m = mask.data();
    for (std::size_t i = 0; i < n; ++i) {
        if (m[i]) out.push_back(p[i]);
    }
    return out;
}

// Scatter `vals` into `out[mask]` (out has the same shape as mask).
// Entries where mask == 0 are left untouched.
void scatter_masked(Array2D<double>& out,
                    const Array2D<std::uint8_t>& mask,
                    const std::vector<double>& vals) {
    const std::size_t n = mask.size();
    const std::uint8_t* m = mask.data();
    double* p = out.data();
    std::size_t k = 0;
    for (std::size_t i = 0; i < n; ++i) {
        if (m[i]) {
            if (k >= vals.size())
                throw std::invalid_argument("scatter_masked: vals shorter than mask count");
            p[i] = vals[k++];
        }
    }
    if (k != vals.size())
        throw std::invalid_argument("scatter_masked: vals longer than mask count");
}

} // namespace

// ---------------------------------------------------------------------------
// acquire_ref kernels
// ---------------------------------------------------------------------------

AcquireRefResult acquire_ref(const Array2D<double>& camlo_ref_im,
                             const Array2D<std::uint8_t>& wfs_mask,
                             const Array2D<double>& dark_im,
                             bool flux_norm) {
    require_same_shape("acquire_ref",
                       camlo_ref_im.rows(), camlo_ref_im.cols(),
                       wfs_mask.rows(), wfs_mask.cols());
    require_same_shape("acquire_ref",
                       camlo_ref_im.rows(), camlo_ref_im.cols(),
                       dark_im.rows(), dark_im.cols());

    AcquireRefResult res;
    res.ref_image = Array2D<double>(camlo_ref_im.rows(), camlo_ref_im.cols(), 0.0);
    const std::size_t n = camlo_ref_im.size();
    for (std::size_t i = 0; i < n; ++i) {
        const double v = (camlo_ref_im.data()[i] - dark_im.data()[i]) * (wfs_mask.data()[i] ? 1.0 : 0.0);
        res.ref_image.data()[i] = v;
    }

    res.flux_norm_coeff = 0.0;
    if (flux_norm) {
        const double s = sum_masked(res.ref_image, wfs_mask);
        if (s == 0.0) {
            throw std::runtime_error("acquire_ref: masked sum is zero, cannot flux-normalize");
        }
        for (std::size_t i = 0; i < n; ++i) {
            res.ref_image.data()[i] /= s;
        }
        res.flux_norm_coeff = s;
    }
    return res;
}

AcquireRefResult acquire_ref(const Array2D<double>& camlo_ref_im,
                             const Array2D<std::uint8_t>& wfs_mask,
                             double dark_scalar,
                             bool flux_norm) {
    Array2D<double> dark(camlo_ref_im.rows(), camlo_ref_im.cols(), dark_scalar);
    return acquire_ref(camlo_ref_im, wfs_mask, dark, flux_norm);
}

// ---------------------------------------------------------------------------
// reconstruct kernel
// ---------------------------------------------------------------------------

std::vector<double> reconstruct(const Array2D<double>& camlo_im,
                                const Array2D<double>& ref_im,
                                const Array2D<std::uint8_t>& wfs_mask,
                                const Array2D<double>& control_matrix,
                                std::size_t mode_lo,
                                std::size_t mode_hi,
                                const Array2D<double>* dark_im,
                                double dark_scalar,
                                bool flux_norm,
                                Array2D<double>* del_im_out) {
    require_same_shape("reconstruct(camlo,ref)",
                       camlo_im.rows(), camlo_im.cols(),
                       ref_im.rows(), ref_im.cols());
    require_same_shape("reconstruct(camlo,mask)",
                       camlo_im.rows(), camlo_im.cols(),
                       wfs_mask.rows(), wfs_mask.cols());

    const std::size_t Nmask = count_mask(wfs_mask);
    if (control_matrix.cols() != Nmask) {
        throw std::invalid_argument(
            "reconstruct: control_matrix has " + std::to_string(control_matrix.cols())
            + " columns but wfs_mask has " + std::to_string(Nmask) + " active pixels");
    }
    if (mode_hi > control_matrix.rows() || mode_lo > mode_hi) {
        throw std::invalid_argument(
            "reconstruct: mode range [" + std::to_string(mode_lo)
            + "," + std::to_string(mode_hi) + ") out of bounds for "
            + std::to_string(control_matrix.rows()) + " modes");
    }

    // 1) dark-subtract (in-place into a working buffer)
    Array2D<double> work(camlo_im.rows(), camlo_im.cols(), 0.0);
    const std::size_t n = camlo_im.size();
    if (dark_im) {
        require_same_shape("reconstruct(dark)",
                           camlo_im.rows(), camlo_im.cols(),
                           dark_im->rows(), dark_im->cols());
        for (std::size_t i = 0; i < n; ++i) {
            work.data()[i] = camlo_im.data()[i] - dark_im->data()[i];
        }
    } else {
        for (std::size_t i = 0; i < n; ++i) {
            work.data()[i] = camlo_im.data()[i] - dark_scalar;
        }
    }

    // 2) optional flux normalization by the masked sum of the dark-subtracted image
    if (flux_norm) {
        const double s = sum_masked(work, wfs_mask);
        if (s != 0.0) {
            for (std::size_t i = 0; i < n; ++i) work.data()[i] /= s;
        }
    }

    // 3) del_im = camlo_im_flux_norm - ref_im
    for (std::size_t i = 0; i < n; ++i) {
        work.data()[i] -= ref_im.data()[i];
    }
    if (del_im_out) {
        *del_im_out = work; // copy out before continuing
    }

    // 4) Extract masked entries and gemv against the [mode_lo, mode_hi) slice.
    std::vector<double> del_masked = extract_masked(work, wfs_mask);
    std::vector<double> coeff(mode_hi - mode_lo, 0.0);

    // The control matrix is row-major (Nmodes, Nmask). We do
    //     coeff[k] = sum_j  control_matrix[mode_lo+k][j] * del_masked[j]
    // which is gemv with stride = Nmask.
    const double* C = control_matrix.data();
    const std::size_t Ncol = control_matrix.cols();
    for (std::size_t k = 0; k < coeff.size(); ++k) {
        const double* row = C + (mode_lo + k) * Ncol;
        double s = 0.0;
        for (std::size_t j = 0; j < Ncol; ++j) {
            s += row[j] * del_masked[j];
        }
        coeff[k] = s;
    }
    return coeff;
}

// ---------------------------------------------------------------------------
// compute_zpo kernel
// ---------------------------------------------------------------------------

Array2D<double> compute_zpo(const std::vector<std::vector<double>>& dm_commands_masked,
                            const Array2D<std::uint8_t>& wfs_mask,
                            const Array2D<double>& response_matrix,
                            const Array2D<double>& dm_modal_matrix) {
    const std::size_t Nmask = count_mask(wfs_mask);
    if (response_matrix.rows() != Nmask) {
        throw std::invalid_argument(
            "compute_zpo: response_matrix has " + std::to_string(response_matrix.rows())
            + " rows but wfs_mask has " + std::to_string(Nmask) + " active pixels");
    }
    const std::size_t Nmodes = response_matrix.cols();
    if (dm_modal_matrix.rows() != Nmodes) {
        throw std::invalid_argument(
            "compute_zpo: dm_modal_matrix.rows (" + std::to_string(dm_modal_matrix.rows())
            + ") != response_matrix.cols (" + std::to_string(Nmodes) + ")");
    }
    const std::size_t Ndm = dm_modal_matrix.cols();

    std::vector<double> zpo_masked(Nmask, 0.0);
    std::vector<double> tmp_modes(Nmodes, 0.0);
    std::vector<double> tmp_pixels(Nmask, 0.0);

    for (const auto& dm_cmd : dm_commands_masked) {
        if (dm_cmd.size() != Ndm) {
            throw std::invalid_argument(
                "compute_zpo: dm_command size " + std::to_string(dm_cmd.size())
                + " != dm_modal_matrix.cols " + std::to_string(Ndm));
        }

        // tmp_modes = dm_modal_matrix . dm_cmd   (Nmodes x Ndm) * (Ndm) -> (Nmodes)
        const double* M = dm_modal_matrix.data();
        for (std::size_t i = 0; i < Nmodes; ++i) {
            const double* row = M + i * Ndm;
            double s = 0.0;
            for (std::size_t j = 0; j < Ndm; ++j) s += row[j] * dm_cmd[j];
            tmp_modes[i] = s;
        }

        // tmp_pixels = response_matrix . tmp_modes  (Nmask x Nmodes) * (Nmodes) -> (Nmask)
        const double* R = response_matrix.data();
        for (std::size_t i = 0; i < Nmask; ++i) {
            const double* row = R + i * Nmodes;
            double s = 0.0;
            for (std::size_t j = 0; j < Nmodes; ++j) s += row[j] * tmp_modes[j];
            tmp_pixels[i] = s;
        }

        // zpo_masked += tmp_pixels
        for (std::size_t i = 0; i < Nmask; ++i) zpo_masked[i] += tmp_pixels[i];
    }

    Array2D<double> zpo(wfs_mask.rows(), wfs_mask.cols(), 0.0);
    scatter_masked(zpo, wfs_mask, zpo_masked);
    return zpo;
}

// ---------------------------------------------------------------------------
// loop_step kernel: full per-iteration math
// ---------------------------------------------------------------------------

Array2D<double> loop_step(const Array2D<double>& camlo_im,
                          const Array2D<double>& ref_plus_zpo,
                          const Array2D<std::uint8_t>& wfs_mask,
                          const Array2D<double>& control_matrix,
                          const Array2D<double>& dm_modes_flat,
                          std::size_t dm_rows,
                          std::size_t dm_cols,
                          const std::vector<double>& gains,
                          std::size_t mode_lo,
                          std::size_t mode_hi,
                          const std::vector<double>& ffo,
                          const Array2D<double>* dark_im,
                          double dark_scalar,
                          bool flux_norm) {
    const std::size_t Nmodes_active = mode_hi - mode_lo;
    if (gains.size() < Nmodes_active) {
        throw std::invalid_argument("loop_step: gains shorter than active mode range");
    }
    if (ffo.size() != Nmodes_active && !ffo.empty()) {
        throw std::invalid_argument("loop_step: ffo length must match (mode_hi - mode_lo) or be empty");
    }
    if (dm_modes_flat.cols() != dm_rows * dm_cols) {
        throw std::invalid_argument("loop_step: dm_modes_flat.cols != dm_rows*dm_cols");
    }
    if (dm_modes_flat.rows() < mode_hi) {
        throw std::invalid_argument("loop_step: dm_modes_flat has fewer rows than mode_hi");
    }

    // 1) reconstruct
    std::vector<double> coeff = reconstruct(
        camlo_im, ref_plus_zpo, wfs_mask, control_matrix,
        mode_lo, mode_hi, dark_im, dark_scalar, flux_norm);

    // 2) coeff -= ffo (when provided)
    if (!ffo.empty()) {
        for (std::size_t i = 0; i < coeff.size(); ++i) coeff[i] -= ffo[i];
    }

    // 3) modal_coeff = - gains * coeff (elementwise; gains is indexed
    //    relative to mode_lo so gains[0] applies to mode mode_lo.)
    std::vector<double> modal_coeff(coeff.size());
    for (std::size_t i = 0; i < coeff.size(); ++i) {
        modal_coeff[i] = -gains[i] * coeff[i];
    }

    // 4) del_dm_command = sum_i modal_coeff[i] * dm_modes_flat[mode_lo+i, :]
    Array2D<double> del_dm(dm_rows, dm_cols, 0.0);
    const std::size_t Ndm = dm_rows * dm_cols;
    const double* MFlat = dm_modes_flat.data();
    double* dst = del_dm.data();
    for (std::size_t i = 0; i < modal_coeff.size(); ++i) {
        const double w = modal_coeff[i];
        if (w == 0.0) continue;
        const double* row = MFlat + (mode_lo + i) * Ndm;
        for (std::size_t k = 0; k < Ndm; ++k) dst[k] += w * row[k];
    }
    return del_dm;
}

} // namespace lina::llowfsc
